x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -381,7 +472,42 @@ public void onResponse(MultiGetResponse multiGetItemResponses) {
                     }
                 }
             }
-            if (this.remainingSlices.decrementAndGet() == 0) {
+            mayFinish();
+        }
+
+        public void onHostsResponse(SearchResponse searchResponse) {
+            SearchHit[] hits = searchResponse.getHits().getHits();
+            for (SearchHit hit : hits) {
+                HostMetadata host = HostMetadata.fromSource(hit.getSourceAsMap());
+                if (host != null) {
+                    hostsTable.putIfAbsent(host.hostID, host);
+                }
+            }
+            log.debug(hostsWatch::report);
+            log.debug("Got [{}] host metadata items", hostsTable.size());
+
+            // Do the CO2 and cost calculation in parallel to waiting for frame metadata.
+            StopWatch watch = new StopWatch("calculateCO2AndCosts");
+            CO2Calculator co2Calculator = new CO2Calculator(costsService, hostsTable, responseBuilder.requestedDuration);
+            CostCalculator costCalculator = new CostCalculator(costsService, hostsTable, responseBuilder.requestedDuration);
+            Map<String, TraceEvent> events = responseBuilder.stackTraceEvents;
+            for (HostEventCount hec : responseBuilder.hostEventCounts) {
+                TraceEvent event = events.get(hec.stacktraceID);
+                if (event == null) {
+                    // this should never happen
+                    log.error("CO2/cost calculator: missing trace events for StackTraceID " + hec.stacktraceID);
```

Comment:
Should this lead to an exception and thus an error response instead? If not, we should not log this on error level but rather as warning. Also, Elasticsearch wraps arguments in square brackets in log statements, i.e:

```
log.warning("CO2/cost calculator: missing trace events for StackTraceID [" + hec.stacktraceID + "].");
```

Finally, it might make sense to collect problematic stack trace ids and log them collectively at the end (you can use `Strings#collectionToDelimitedStringWithLimit()` to avoid logging too many ids without adding additional value).